### PR TITLE
Feature/fts index fix

### PR DESCRIPTION
**Pull request by hajnalandor**

### CODE REVIEW

File: fts_index.go [hunk 1]
```diff
@@ -194,6 +225,25 @@ func (h *Handler) CreateFullTextSearchIndex(ctx context.Context, def *IndexDefin
 	return nil
 }
 
+//func (h *Handler) CreateDefaultFullTextSearchIndex(ctx context.Context, indexName string, indexTypes map[string]IndexType) error {
```
Comment (borosr) [inline fts_index.go]: Remove this

File: testing.go [hunk 2]
```diff
@@ -33,6 +33,12 @@ func seed() {
 	th.SetDocumentType(context.Background(), "product", "product")
 	th.SetDocumentType(context.Background(), "store", "store")
 
+	createFullTextSearchIndex("webshop_fts_index", false, "webshop")
+	createFullTextSearchIndex("product_fts_index", false, "product")
+	createFullTextSearchIndex("store_fts_index", false, "store")
+	createFullTextSearchIndex("order_fts_idx", false, "order")
```
Comment (borosr) [inline testing.go]: order_fts_idx is now order_fts_index

File: testing.go [hunk 3]
```diff
@@ -33,6 +33,12 @@ func seed() {
 	th.SetDocumentType(context.Background(), "product", "product")
 	th.SetDocumentType(context.Background(), "store", "store")
 
+	createFullTextSearchIndex("webshop_fts_index", false, "webshop")
```
Comment (borosr) [inline testing.go]: These indexes everything?

**Comment by PumpkinSeed** [comment]
#71

File: testing.go (hunk 2 above)
Comment (borosr) [reply testing.go]: order_fts_idx is used

**Comment by PumpkinSeed** [comment]
https://docs.couchbase.com/server/6.0/fts/fts-creating-indexes.html

**Comment by borosr** [comment]
https://forums.couchbase.com/t/full-text-search-index-progress-status/23640/2

File: testing.go [hunk 4]
```diff
@@ -212,3 +248,61 @@ func createFullTextSearchIndex(indexName string, deleteOnExists bool) error {
...
+					time.Sleep(10 * time.Millisecond)
+				} else {
+					break
+				}
+			}
+		}
+	}
+
+	// NOTE: Sleep because most of the tests want to use this index, so it should wait for
+	//time.Sleep(5 * time.Second)
```
Comment (PumpkinSeed) [inline testing.go]: Remove it

File: testing.go [hunk 5]
```diff
@@ -51,6 +57,7 @@ func seed() {
 		fmt.Printf("Connection setup, data seeded %v\n", time.Since(start))
 		seeded = true
 	}
+	time.Sleep(5 * time.Second)
```
Comment (PumpkinSeed) [inline testing.go]: is it necessary?

File: testing.go [hunk 6]
```diff
@@ -33,6 +33,12 @@ func seed() {
 	th.SetDocumentType(context.Background(), "product", "product")
 	th.SetDocumentType(context.Background(), "store", "store")
 
+	createFullTextSearchIndex("webshop_fts_index", false, "webshop")
+	createFullTextSearchIndex("product_fts_index", false, "product")
+	createFullTextSearchIndex("store_fts_index", false, "store")
+	createFullTextSearchIndex("order_fts_idx", false, "order")
+	time.Sleep(1 * time.Second)
```
Comment (PumpkinSeed) [inline testing.go]: is it necessary?

File: indexer_test.go [hunk 7]
```diff
@@ -42,6 +43,7 @@ func TestPrimaryIndexCreateError(t *testing.T) {
 }
 
 func TestExistingIndex(t *testing.T) {
+	// @TODO hanging
```
Comment (PumpkinSeed) [inline indexer_test.go]: hanging?

File: indexer_test.go [hunk 8]
```diff
@@ -9,6 +9,7 @@ import (
 )
 
 func TestIndexCreate(t *testing.T) {
+	// @TODO hanging
```
Comment (PumpkinSeed) [inline indexer_test.go]: hanging?

File: fts_index_test.go [hunk 9]
```diff
@@ -48,3 +53,201 @@ func TestDefaultFullTextSearchIndexDefinitionWithoutSourceName(t *testing.T) {
...
+			},
+		},
+	}
+
+	if err := th.CreateFullTextSearchIndex(context.Background(), def); err != nil {
+		assert.Nil(t, th.DeleteFullTextSearchIndex(context.Background(), indexName))
+		assert.Nil(t, th.CreateFullTextSearchIndex(context.Background(), def))
+	}
+
+	time.Sleep(2 * time.Second)
```
Comment (PumpkinSeed) [inline fts_index_test.go]: is it necessary?

File: testing.go [hunk 10]
```diff
@@ -212,3 +248,61 @@ func createFullTextSearchIndex(indexName string, deleteOnExists bool) error {
...
+		}
+		if err != nil {
+			return err
+		}
+		err = th.CreateFullTextSearchIndex(ctx, def)
+		if err != nil {
+			return err
+		}
+
+		for true {
```
Comment (PumpkinSeed) [inline testing.go]: Remove true, for {} will be an endless loop

File: testing.go [hunk 11]
```diff
@@ -212,3 +248,61 @@ func createFullTextSearchIndex(indexName string, deleteOnExists bool) error {
...
+			count, _ := th.countIndex(ctx, indexName)
+			stat, _ := th.indexStat(ctx, indexName)
+			if !count.Count.Valid || !stat.DocCount.Valid {
+				log.Printf("%s count[%v] or stat invalid[%v]\n", indexName, count.Count.Valid, stat.DocCount.Valid)
```
Comment (PumpkinSeed) [inline testing.go]: don't log it, or use user-defined logger

File: testing.go [hunk 12]
```diff
@@ -212,3 +248,61 @@ func createFullTextSearchIndex(indexName string, deleteOnExists bool) error {
...
+				time.Sleep(10 * time.Millisecond)
+			}
+			if count.Count.Uint > 0 {
+				if stat.DocCount.Uint != count.Count.Uint {
+					log.Printf("%s count is %d\n stat is %d\n", indexName, count.Count.Uint, stat.DocCount.Uint)
```
Comment (PumpkinSeed) [inline testing.go]: don't log it, or use user-defined logger

File: fts_index.go [hunk 13]
```diff
@@ -257,3 +308,49 @@ func (h *Handler) fullTextSearchURL(ctx context.Context, indexName string) strin
 	}
 	return fmt.Sprintf("%s%s/%s", h.httpAddress, ftsEndpoint, indexName)
 }
+
+func (h *Handler) countIndex(ctx context.Context, indexName string) (*IndexCount, error) {
```
Comment (PumpkinSeed) [inline fts_index.go]: Export this function

File: fts_index.go [hunk 14]
```diff
@@ -257,3 +308,49 @@ func (h *Handler) fullTextSearchURL(ctx context.Context, indexName string) strin
...
+	}
+	var indexCount IndexCount
+	if err := json.Unmarshal(respbody, &indexCount); err != nil {
+		return nil, err
+	}
+
+	return &indexCount, nil
+}
+
+func (h *Handler) indexStat(ctx context.Context, indexName string) (*IndexStat, error) {
```
Comment (PumpkinSeed) [inline fts_index.go]: Export this function

**Pull request merged**